### ISSUE TITLE
Fix clippy warnings

### ISSUE THREAD

**Issue by sunfishcode** [CraneStation/cranelift#392]
Between Cranelift changes and Clippy changes, we appear to have accumulated a number of Clippy warnings again :-). These are often good first issues for people to jump in and fix. Also, it isn't necessary to fix all of them at once.

To run Clippy on Cranelift, install clippy: `rustup component add clippy-preview` (see [here](https://github.com/rust-lang-nursery/rust-clippy) for more info), and then run `cargo clippy`. You can do this with either stable or nightly Rust, though nightly tends to report more warnings.

Sometimes clippy's warnings and suggestions aren't appropriate, so keep in mind all the options:
 - Follow clippy's suggestion
 - Rearrange the code some other way to avoid the warning
 - Add a local `allow`, like this: https://github.com/cretonne/cretonne/blob/master/lib/codegen/src/bforest/pool.rs#L64
 - Add a crate-wide `allow` to the containing crate's lib.rs, like this: https://github.com/cretonne/cretonne/blob/master/lib/codegen/src/lib.rs#L7

and either make a judgement call, or ask here in this issue.

**Comment by Aaronepower** [comment]
@sunfishcode How should the following lints about whether a argument should be passed by reference or value be handled?

```

warning: this argument is passed by value, but not consumed in the function body
   --> lib/codegen/src/flowgraph.rs:163:34
    |
163 |     fn add_edge(&mut self, from: BasicBlock, to: Ebb) {
    |                                  ^^^^^^^^^^ help: consider taking a reference instead: `&BasicBlock`
    |
    = note: #[warn(needless_pass_by_value)] on by default
help: consider marking this type as Copy
   --> lib/codegen/src/flowgraph.rs:35:1
    |
35  | / pub struct BasicBlock {
36  | |     /// Enclosing Ebb key.
37  | |     pub ebb: Ebb,
38  | |     /// Last instruction in the basic block.
39  | |     pub inst: Inst,
40  | | }
    | |_^
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#needless_pass_by_value

warning: this argument is passed by reference, but would be more efficient if passed by value
   --> lib/codegen/src/ir/instructions.rs:450:21
    |
450 |     fn is_base_type(&self, scalar: Type) -> bool {
    |                     ^^^^^ help: consider passing by value instead: `self`
    |
    = note: #[warn(trivially_copy_pass_by_ref)] on by default
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#trivially_copy_pass_by_ref

warning: this argument is passed by reference, but would be more efficient if passed by value
   --> lib/codegen/src/ir/instructions.rs:464:21
    |
464 |     pub fn contains(&self, typ: Type) -> bool {
    |                     ^^^^^ help: consider passing by value instead: `self`
    |
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#trivially_copy_pass_by_ref

warning: this argument is passed by reference, but would be more efficient if passed by value
   --> lib/codegen/src/ir/instructions.rs:472:20
    |
472 |     pub fn example(&self) -> Type {
    |                    ^^^^^ help: consider passing by value instead: `self`
    |
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#trivially_copy_pass_by_ref

warning: this argument is passed by reference, but would be more efficient if passed by value
  --> lib/codegen/src/ir/valueloc.rs:29:24
   |
29 |     pub fn is_assigned(&self) -> bool {
   |                        ^^^^^ help: consider passing by value instead: `self`
   |
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#trivially_copy_pass_by_ref

warning: this argument is passed by reference, but would be more efficient if passed by value
   --> lib/codegen/src/ir/valueloc.rs:114:24
    |
114 |     pub fn is_assigned(&self) -> bool {
    |                        ^^^^^ help: consider passing by value instead: `self`
    |
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#trivially_copy_pass_by_ref

warning: this argument is passed by reference, but would be more efficient if passed by value
   --> lib/codegen/src/ir/valueloc.rs:122:19
    |
122 |     pub fn is_reg(&self) -> bool {
    |                   ^^^^^ help: consider passing by value instead: `self`
    |
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#trivially_copy_pass_by_ref

warning: this argument is passed by reference, but would be more efficient if passed by value
   --> lib/codegen/src/ir/valueloc.rs:130:21
    |
130 |     pub fn is_stack(&self) -> bool {
    |                     ^^^^^ help: consider passing by value instead: `self`
    |
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#trivially_copy_pass_by_ref

warning: this argument is passed by value, but not consumed in the function body
 --> /src/rust/cranelift/target/debug/build/cranelift-codegen-3ef500751b344055/out/settings-riscv.rs:9:51
  |
9 |     pub fn new(shared: &settings::Flags, builder: Builder) -> Self {
  |                                                   ^^^^^^^ help: consider taking a reference instead: `&Builder`
  |
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#needless_pass_by_value

warning: this argument is passed by value, but not consumed in the function body
 --> /src/rust/cranelift/target/debug/build/cranelift-codegen-3ef500751b344055/out/settings-x86.rs:9:51
  |
9 |     pub fn new(shared: &settings::Flags, builder: Builder) -> Self {
  |                                                   ^^^^^^^ help: consider taking a reference instead: `&Builder`
  |
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#needless_pass_by_value

warning: this argument is passed by value, but not consumed in the function body
 --> /src/rust/cranelift/target/debug/build/cranelift-codegen-3ef500751b344055/out/settings-arm32.rs:9:51
  |
9 |     pub fn new(shared: &settings::Flags, builder: Builder) -> Self {
  |                                                   ^^^^^^^ help: consider taking a reference instead: `&Builder`
  |
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#needless_pass_by_value

warning: this argument is passed by value, but not consumed in the function body
 --> /src/rust/cranelift/target/debug/build/cranelift-codegen-3ef500751b344055/out/settings-arm64.rs:9:51
  |
9 |     pub fn new(shared: &settings::Flags, builder: Builder) -> Self {
  |                                                   ^^^^^^^ help: consider taking a reference instead: `&Builder`
  |
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#needless_pass_by_value

warning: this argument is passed by reference, but would be more efficient if passed by value
   --> lib/codegen/src/settings.rs:304:26
    |
304 |         pub fn is_preset(&self) -> bool {
    |                          ^^^^^ help: consider passing by value instead: `self`
    |
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#trivially_copy_pass_by_ref

warning: this argument is passed by value, but not consumed in the function body
 --> /src/rust/cranelift/target/debug/build/cranelift-codegen-3ef500751b344055/out/settings.rs:9:25
  |
9 |     pub fn new(builder: Builder) -> Self {
  |                         ^^^^^^^ help: consider taking a reference instead: `&Builder`
  |
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#needless_pass_by_value
```

**Comment by sunfishcode** [comment]
For the `BasicBlock` one, I think we could change `add_edge` to take two arguments, an `Ebb` and an `Inst`, rather than taking a `BasicBlock` of them, since all `add_edge` does is pull out the parts anyway.

For the trivially_copy_pass_by_ref ones, clippy's suggestion of removing the '&'s sounds good.

For the `Builder` ones, would it work to make the [`state_for`](https://github.com/CraneStation/cranelift/blob/master/lib/codegen/src/settings.rs#L61) function take a `self` rather than a `&self`, and return the `Box<[u8]>`, so that it consumes the `Builder`? I think that would more clearly reflect how it's being used.

**Comment by sunfishcode** [comment]
Note that the `subsec_nanos() / 1_000_000` issue, which is warned about by the `subsec_millis` lint, is something we can't yet fix, because we still support Rust 1.25, and `subsec_millis` is new. We should instead add this as an `allow` for now.

**Comment by waywardmonkeys** [comment]
I've updated my PR (#408) for the `subsec_nanos` thing.
